### PR TITLE
Add support to forward external domains DNS queries to dnsmasq

### PR DESCRIPTION
**Pull request by danpawlik**
That feature will be mostly used by the CI for simulating the external domains that exists normaly in the Internet.
For example, if domain '.sftests.com' have many subdomains and not each domain is registered domain in the Openshift cluster and the domain is just for test (not registered in any DNS authority) it will fail name resolution inside the pod with an error:

    "curl: (6) Could not resolve host: test.sftests.com"

In that case, we are adding custom domain inside the codedns configuration configmap, which will forward the DNS queries made for that domain to the dnsmasq, which is installed on the host. The dnsmasq (which is started by NetworkManager) have proper entries to resolve the domain.

### CODE REVIEW

File: defaults/main.yaml [hunk 1]
```diff
@@ -142,3 +142,13 @@ public_dns:
...
 # Set the cri-o registry policy to pull images even from untrasted registries.
 overwrite_container_policy: false
+
+# Set additional domains, that should be resolved in pods, but you don't
+# want to use Kubernetes like domain resolution. It might be helpful for CI
+# tests.
+# For example:
+# additional_domains:
+#   - name: sftests.com
+#     address: 10.0.0.1
```
Comment (TristanCacqueray) [inline defaults/main.yaml:152]: I thought that thanks to dnsmasq, the route should already resolve on the host, so why do we need to define the domain explicitly?

File: tasks/additional_domains.yaml [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        yq ".data.Corefile = \"$(cat {{ coredns_config_dir }}/corefile.raw)\"" /etc/microshift/dns-configmap.yaml > {{ coredns_config_dir }}/patched-corefile.yaml
+
+    - name: Remove keys that will block coredns configmap update
+      ansible.builtin.shell: |
+        sed -i '/  creationTimestamp:\|  resourceVersion:\|  uid:/d' {{ coredns_config_dir }}/patched-corefile.yaml
+
+    - name: Apply new coredns configmap
+      ansible.builtin.shell: |
+        kubectl apply -f {{ coredns_config_dir }}/patched-corefile.yaml
+      notify: Restart Openshift DNS
```
Comment (TristanCacqueray) [inline tasks/additional_domains.yaml:67]: Isn't there a simpler way to update the config map? These tasks looks a bit brittle. Isn't there a kubectl invocation to inject the new Corefile in the cm?

**Review by TristanCacqueray — CHANGES_REQUESTED** [review]

File: templates/corefile.yaml [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    }
+    ready
+    kubernetes cluster.local in-addr.arpa ip6.arpa {
+        pods insecure
+        fallthrough in-addr.arpa ip6.arpa
+    }
+    prometheus 127.0.0.1:9153
+    forward . /etc/resolv.conf {
+        policy sequential
+    }
```
Comment (TristanCacqueray) [inline templates/corefile.yaml:18]: Why not simply:  `forward . {{ ansible_ipv4 }}:53`

File: defaults/main.yaml (hunk 1 above)
Comment (danpawlik) [reply defaults/main.yaml:152]: that's a good point.
After last debuging, some part of the PS would be changed.
tl;dr by setting dns=dnsmasq in NetworkManager, the nameserver is set to 127.0.0.1 and after restarting the openshuft-dns daemonset, pods are trying to resolve domains by making query to 127.0.0.1. Soon I will do that patch set ready to review!

File: defaults/main.yaml [hunk 4]
```diff
@@ -140,5 +140,8 @@ public_dns:
   - 8.8.8.8
   - 9.9.9.9
 
+# A place where openshift-dns corefile will be stored and changed.
+coredns_config_dir: "/etc/coredns-configs"
```
Comment (TristanCacqueray) [inline defaults/main.yaml:144]: Why are we keeping a copy outside of the config map?

File: templates/corefile.yaml [hunk 5]
```diff
@@ -0,0 +1,29 @@
+.:5353 {
```
Comment (TristanCacqueray) [inline templates/corefile.yaml:1]: Is this yaml ?

File: tasks/coredns.yaml [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+    - name: Remove keys that will block coredns configmap update
+      become: true
+      ansible.builtin.shell: |
+        sed -i '/  creationTimestamp:\|  resourceVersion:\|  uid:/d' {{ coredns_config_dir }}/patched-corefile.yaml
+
+    - name: Apply new coredns configmap
+      ansible.builtin.shell: |
+        kubectl apply -f {{ coredns_config_dir }}/patched-corefile.yaml
+      notify: Restart Openshift DNS
```
Comment (TristanCacqueray) [inline tasks/coredns.yaml:77]: Is there a reason we are not using the kubectl command to update the config, e.g.:

```
    kubectl create cm dns-default --from-file=./Corefile --dry-run -o yaml | \
      kubectl -n openshift-dns apply -f -
```

**Review by TristanCacqueray — CHANGES_REQUESTED** [review]

File: templates/corefile.yaml (hunk 5 above)
Comment (danpawlik) [reply templates/corefile.yaml:1]: no, it is like json, but withou quotes

File: tasks/coredns.yaml (hunk 6 above)
Comment (danpawlik) [reply tasks/coredns.yaml:77]: yes, such update is sensitive IMHO and would like to keep all the files that have been created. 
Normally, it can be done simple by using kubectl patch + /usr/local/bin/yq ".data.Corefile = \"$(cat {{ coredns_config_dir }}/corefile.raw)\"", but it looks ugly and was harder to find a bug

File: defaults/main.yaml (hunk 4 above)
Comment (danpawlik) [reply defaults/main.yaml:144]: IMHO It's better to keep a backup. Also later we are operating on the default config

**Review by TristanCacqueray — APPROVED** [review]
Thanks!

File: templates/corefile.yaml (hunk 5 above)
Comment (TristanCacqueray) [reply templates/corefile.yaml:1]: Ok, then perhaps the file should be named `Corefile.j2`

File: tasks/coredns.yaml (hunk 6 above)
Comment (TristanCacqueray) [reply tasks/coredns.yaml:77]: I see, that's annoying. Well perhaps we could dump the whole cm, and use the `kubectl create --dry-run | kubectl apply` trick to update the Corefile
